ip-utils.c

```diff
@@ -0,0 +1,159 @@
+/*! \file    ip-utils.c
+ * \author   Johan Ouwerkerk <jm.ouwerkerk@gmail.com>
+ * \copyright GNU General Public License v3
+ * \brief    IP address related utility functions (headers)
+ * \details  Provides functions to query for network devices with a given device name or address.
+ * Devices may be looked up by either a device name or by the IPv4 or IPv6 address of the configured network interface.
+ * This functionality may be used to bind to user configurable network devices instead of relying on unpredictable implementation defined defaults.
+ * Parsing IPv4/IPv6 addresses is done using inet_pton(), making these functions generally robust against malformed input.
+ *
+ * \ingroup core
+ * \ref core
+ */
+
+#include <errno.h>
+#include <sys/types.h>
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <string.h>
+
+#include "ip-utils.h"
+
+static int compare_byte_arrays(const uint8_t * b1, const uint8_t * b2, const size_t size) {
+	size_t i;
+	for(i = 0; i < size; ++i) {
+		if(b1[i] != b2[i]) {
+			return b1[i] > b2[i] ? 1 : -1;
+		}
+	}
+	return 0;
+}
+
+static int iface_matches_name(const struct ifaddrs * ifa, const char * name) {
+	return name && (ifa->ifa_addr->sa_family == AF_INET || ifa->ifa_addr->sa_family == AF_INET6) && strncmp(ifa->ifa_name, name, IFNAMSIZ) == 0 ? 1 : 0;
+}
+
+static int iface_matches_ipv4(const struct ifaddrs * ifa, const struct in_addr * ipv4) {
+	if(ifa->ifa_addr->sa_family == AF_INET) {
+		struct sockaddr_in * iface = (struct sockaddr_in *) ifa->ifa_addr;
+		if(iface->sin_addr.s_addr == ipv4->s_addr) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static int iface_matches_ipv6(const struct ifaddrs * ifa, const struct in6_addr * ipv6) {
+	if(ifa->ifa_addr->sa_family == AF_INET6) {
+		struct sockaddr_in6 * iface = (struct sockaddr_in6 *) ifa->ifa_addr;
+		if(compare_byte_arrays(iface->sin6_addr.s6_addr, ipv6->s6_addr, 16) == 0) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static int iface_matches(const struct ifaddrs * ifa, const janus_network_query_config * query) {
+	if(ifa && query) {
+		return
+			((query->mode & IPv4OnlyMatch) && iface_matches_ipv4(ifa, &query->ipv4) == 1) ||
+			((query->mode & IPv6OnlyMatch) && iface_matches_ipv6(ifa, &query->ipv6) == 1) ||
+			((query->mode & NameOnlyMatch) && iface_matches_name(ifa, query->device_name) == 1) ? 1 : 0;
+	}
+	else {
+		return -EINVAL;
+	}
+}
+
+const struct ifaddrs * janus_query_network_devices(const struct ifaddrs * ifa, const janus_network_query_config * query) {
```

Comment:
Extra space after the asterisks (per coding style, we write pointers like `type *pointer`). Please fix wherever it applies.